gitclone/bitriseapi/pr_merge_ref.go

```diff
@@ -0,0 +1,136 @@
+package bitriseapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/bitrise-io/go-utils/retry"
+	"github.com/bitrise-io/go-utils/v2/log"
+	"github.com/bitrise-steplib/steps-git-clone/gitclone/tracker"
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+const maxAttemptCount = 5
+const retryWaitTime = time.Second * 2
+
+type MergeRefChecker interface {
+	// IsMergeRefUpToDate returns true if the ref is safe to use in a checkout, and it reflects the latest state of the PR
+	IsMergeRefUpToDate(ref string) (bool, error)
+}
+
+func NewMergeRefChecker(buildURL string, apiToken string, client *retryablehttp.Client, logger log.Logger, tracker tracker.StepTracker) MergeRefChecker {
+	return apiMergeRefChecker{
+		buildURL: buildURL,
+		apiToken: apiToken,
+		client:   client,
+		logger:   logger,
+		tracker:  tracker,
+	}
+}
+
+type apiMergeRefChecker struct {
+	buildURL string
+	apiToken string
+	client   *retryablehttp.Client
+	logger   log.Logger
+	tracker  tracker.StepTracker
+}
+
+type mergeRefResponse struct {
+	Status string `json:"status"`
+	Error  string `json:"error_msg"`
+}
+
+type mergeRefFetcher func(attempt uint) (mergeRefResponse, error)
+
+func (c apiMergeRefChecker) IsMergeRefUpToDate(ref string) (bool, error) {
+	if c.buildURL == "" {
```

Comment:
I made this change because it seemed like a good idea, then I realized why we don't do this 😅 

This `MergeRefChecker` is instantiated at the beginning of the main execution, so handling the constructor errors would happen there and would abort the entire step run. This is exactly what happened in the E2E tests after my changes, a perfectly fine test failed because it didn't mock the API URL and token, but there is no API communication in that test case.

Since this merge ref checker is just one of the many code paths and we handle its errors with a fallback anyway, I think it's better to leave it as-is, I don't want to make the local step execution (or E2E execution) require an API token.